bundles/org.openhab.binding.sonos/src/main/java/org/openhab/binding/sonos/internal/handler/ZonePlayerHandler.java

```diff
@@ -1600,6 +1667,79 @@ public void setVolumeForGroup(Command command) {
+                inputs.put("InstanceID", "0");
+                inputs.put("DesiredBass", newValue);
+
+                Map<String, String> result = service.invokeAction(this, "RenderingControl", "SetBass", inputs);
+
+                for (String variable : result.keySet()) {
+                    this.onValueReceived(variable, result.get(variable), "RenderingControl");
+                }
+            }
+        }
+    }
+
+    public void setTreble(Command command) {
+        if (!isOutputLevelFixed()) {
+            String newValue = getNewNumericValue(command, getTreble(), MIN_TREBLE, MAX_TREBLE);
+            if (newValue != null) {
+                Map<String, String> inputs = new HashMap<>();
+                inputs.put("InstanceID", "0");
+                inputs.put("DesiredTreble", newValue);
+
+                Map<String, String> result = service.invokeAction(this, "RenderingControl", "SetTreble", inputs);
+
+                for (String variable : result.keySet()) {
+                    this.onValueReceived(variable, result.get(variable), "RenderingControl");
+                }
+            }
+        }
+    }
+
+    private @Nullable String getNewNumericValue(Command command, @Nullable String currentValue, int minValue,
+            int maxValue) {
+        String newValue = null;
+        if (command instanceof IncreaseDecreaseType || command instanceof DecimalType) {
+            if (command == IncreaseDecreaseType.INCREASE && currentValue != null) {
+                int i = Integer.valueOf(currentValue);
+                newValue = String.valueOf(Math.min(maxValue, i + 1));
+            } else if (command == IncreaseDecreaseType.DECREASE && currentValue != null) {
+                int i = Integer.valueOf(currentValue);
+                newValue = String.valueOf(Math.max(minValue, i - 1));
+            } else if (command instanceof DecimalType) {
+                newValue = String.valueOf(((DecimalType) command).intValue());
+            }
+        }
+        return newValue;
+    }
+
+    public void setLoudness(Command command) {
+        if (!isOutputLevelFixed() && (command instanceof OnOffType || command instanceof OpenClosedType
+                || command instanceof UpDownType)) {
+            Map<String, String> inputs = new HashMap<>();
+            inputs.put("InstanceID", "0");
+            inputs.put("Channel", "Master");
+
+            if (command.equals(OnOffType.ON) || command.equals(UpDownType.UP) || command.equals(OpenClosedType.OPEN)) {
+                inputs.put("DesiredLoudness", "True");
+            } else if (command.equals(OnOffType.OFF) || command.equals(UpDownType.DOWN)
+                    || command.equals(OpenClosedType.CLOSED)) {
+                inputs.put("DesiredLoudness", "False");
+            }
+
+            Map<String, String> result = service.invokeAction(this, "RenderingControl", "SetLoudness", inputs);
+
+            for (String variable : result.keySet()) {
```

Comment:
Generally better use `entrySet` because `keySet` will create a copy of all keys.
```suggestion
            for (String variable : result.keySet()) {
```

Reply:
Same pattern in a lot of methods of the class. I propose to change in a separate PR.

Reply:
Ok, I finally included the change in this PR and also define constants and factorize some code.